### ISSUE TITLE
`pyright` mishandles optional local function declaration

### ISSUE THREAD

**Issue by ncoghlan** [microsoft/pyright#9114]
**Describe the bug**

`pyright` in the current `pylance` VSCode extension isn't treating local function declarations like normal local variables, so it doesn't infer types correctly when a callable is allowed to be set to `None`.

Previously, when one branch of an `if` statement set the target variable to `None`, and the other defined it as a local function, `pyright` correctly inferred the corresponding union type. Now, even an explicit predeclaration of the union type gets rejected (since pyright is ignoring it in favour of the function signature).

**Code or Screenshots**

`pyright` used to correctly infer a union type for this code, but now errors as noted:

```python
    if max_value is None:
        _clamp_value = None  # pyright complains that None can't be assigned to `(value:int) -> int`
    else:
        def _clamp_value(value:int) -> int:
            return min(max_value, value)
```

Even explicitly declaring the union type doesn't eliminate the problem

```python
    _clamp_value: Callable[[int], int]|None  # pyright complains that this type declaration is overridden
    if max_value is None:
        _clamp_value = None  # pyright complains that None can't be assigned to `(value:int) -> int`
    else:
        def _clamp_value(value:int) -> int:
            return min(max_value, value)
```

Combining the declaration with setting a default value also fails with the same errors (just reported against the one declaration line):

```python
    _clamp_value: Callable[[int], int]|None  = None # Both errors reported here
    if max_value is not None:
        def _clamp_value(value:int) -> int:
            return min(max_value, value)
```

**VS Code extension or command-line**

Running in VSCode via `Pylance`. Checking the pylance extension change log, this should be `pyright` 1.1.378.

There are some release notes for newer versions that seem like they could be potentially relevant to this, so I'll see if update to the current pylance pre-release resolves the problem.

**Comment by ncoghlan** [comment]
The issue is still present after installing the pylance pre-release and reloading the VSCode window:

![image](https://github.com/user-attachments/assets/8a63edf7-aeea-4520-9c02-68dc29e48f8b)

The pre-release contains pyright 1.1.382

**Comment by ncoghlan** [comment]
A similar regression via a slightly different mechanism:

![image](https://github.com/user-attachments/assets/2bae2fd8-1038-43d6-9a02-f744901f7409)

Here, `pyright` is complaining about the declaration of the default progress callback that is used when the function parameter is `None` overriding the original parameter declaration instead of treating it as a type narrowing operation.

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug. This behavior hasn't changed in pyright's type checking logic. I presume that something changed with your pylance configuration. You probably had type checking disabled previously, so these diagnostics were suppressed. If you don't want to see type checking errors in pylance, you can set the typeCheckingMode to "off".

The problem here is that a `def` statement is considered a declaration of a symbol. You are not allowed to redeclare that same symbol with a variable type annotation that doesn't match the declaration provided by the `def` statement, nor are you allowed to assign a value (like `None`) that violates the type declaration.

**Comment by ncoghlan** [comment]
You're right that this was a system config change rather than anything changing in `pyright` (I switched to a new machine recently, and forgot that this feature was turned off on the old machine).

However, it seems strange that the typing spec is loose enough to allow `def` to be handled as overriding an explicit local variable type definition when instantiating any other callable would respect the union declaration, so I started a typing discussion thread about that: https://discuss.python.org/t/expected-behaviour-for-local-function-definitions-in-union-types/66168

**Comment by ncoghlan** [comment]
I'll also note that I'm *not* redefining the symbol. The symbol is already defined as a `Callable|None` union type via a parameter or local variable type annotation.

`pyright` is ignoring that explicit type declaration, and instead saying that `def` can't be used when the target symbol is part of a union.

**Comment by ncoghlan** [comment]
I did work out a reasonably straightforward workaround, which is to introduce a second name to use in the `def` statement. Using the progress callback example:

```python
        def _default_progress_callback(_:Any) -> None:
            pass
        progress_callback = _default_progress_callback
```

The separate assignment statement respects the union declaration.

**Comment by erictraut** [comment]
Please refer to the pyright documentation for an explanation of [type declarations](https://microsoft.github.io/pyright/#/type-concepts).

In your first code sample above, you are declaring the type of the symbol `_clamp_value` via a `def` statement, and you are violating that type declaration when you attempt to assign `None` to this symbol.

In your second code sample above, you are providing two incompatible type declarations for the `_clamp_value` symbol — one with a variable type annotation and a second with a `def` statement.

The typing spec has nothing to say about type declarations or what constitutes a type declaration. This term is unlikely to be added to the typing spec because it would be difficult (probably impossible) to get agreement on the concept.

You can see that [pyright and mypy differ](https://microsoft.github.io/pyright/#/mypy-comparison?id=variable-type-declarations) in their philosophy on this point. I personally think that pyright's behavior is preferable. It results in a cleaner and simpler mental model and eliminates the need for many type annotations that are required in mypy's model.

In any case, pyright is working as intended here. This isn't a bug. This behavior (and the associated mental model that it's based on) is deeply baked in to pyright's logic, so it's unlikely to ever be changed.

**Comment by NeilGirdhar** [comment]
Here's a workaround:
```python
    _clamp_value: None | Callable[[int], int]

    if max_value is None:
        _clamp_value = None
    else:
        def _clamp_value_f(value: int) -> int:
            return min(max_value, value)
        _clamp_value = _clamp_value_f
```

**Comment by NeilGirdhar** [comment]
(Related https://github.com/microsoft/pyright/issues/2846)

**Comment by hauntsaninja** [comment]
I agree that in general, pyright's behaviour surrounding declarations, definitions and narrowing is much better for users than mypy's.

If pyright continues to get user feedback about this, one option if feasible could be to avoid treating `def` as a type declaration specifically only in local scope.

**Comment by Andrej730** [comment]
Also came across this problem and found this issue and indeed it was kind of confusing.
Renaming callback to `_callback` and assigning it to `callback` helps as a workaround.

```python
# pyright: strict
from typing import Callable


def call(value: int, callback: Callable[[int], str] | None): ...


def test(arg: bool):
    # ERROR: Type "None" is not assignable to declared type "(obj: int) -> str"
    callback = None

    if arg:

        def callback(obj: int) -> str:
            a = 25
            b = 45
            c = obj * a * b
            return str(c)

        # Workaround:
        # def _callback(obj: int) -> str:
        #     a = 25
        #     b = 45
        #     c = obj * a * b
        #     return str(c)

        # callback = _callback

    # def callback(obj: int) -> str
    # expected: Callable[[int], str] | None
    reveal_type(callback)
    call(42, callback)
```
